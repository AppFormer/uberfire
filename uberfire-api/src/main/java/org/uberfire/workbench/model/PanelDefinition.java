/*
 * Copyright 2012 JBoss Inc
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package org.uberfire.workbench.model;

import java.util.List;
import java.util.Set;

/**
 * A Panel in the Workbench.
 */
public interface PanelDefinition {

    /**
     * Add a Part to the Panel
     * @param part The Part to add
     */
    public void addPart( final PartDefinition part );

    /**
     * Get the Parts contained in the Panel
     * @return The parts
     */
    public Set<PartDefinition> getParts();

    /**
     * Get all of this Panel's immediate child Panels (i.e. not recursive).
     * @return All children
     */
    public List<PanelDefinition> getChildren();

    /**
     * Insert a Panel as an immediate child at the given Position. If this
     * Panel already has a child at the specified position the existing Panel at
     * that position becomes a child of the Panel being added at the same
     * Position.
     * @param position The Position to add the child
     * @param panel The child Panel
     */
    public void insertChild( final Position position,
                             final PanelDefinition panel );

    /**
     * Append a Panel at the first empty child position. If this Panel already
     * has a child at the specified position the descendants are followed until
     * an empty position is found.
     * @param position The Position to add the child
     * @param panel The child Panel
     */
    public void appendChild( final Position position,
                             final PanelDefinition panel );

    /**
     * Append a Panel at the first empty child position. If this Panel already
     * has a child at the specified position the descendants are followed until
     * an empty position is found.
     * @param panel The child Panel
     */
    public void appendChild(
            final PanelDefinition panel );

    /**
     * Get the Panel's immediate child Panel at the given Position
     * @param position The child Panel's Position
     * @return The child Panel or null, if a child does not exist at the given
     *         Position
     */
    public PanelDefinition getChild( final Position position );

    /**
     * Remove a child from the Panel
     * @param position
     */
    public void removeChild( final Position position );

    /**
     * Is this Panel the root of the Perspective definition
     * @return True if the Panel is the root
     */
    public boolean isRoot();

    /**
     * Get the type of the Panel
     * @return The panel type
     */
    public PanelType getPanelType();

    public PanelType getDefaultChildPanelType();

    /**
     * Get the height of the Panel in pixels
     * @return The height, or null if not set
     */
    public Integer getHeight();

    /**
     * Set the height of the Panel in pixels
     * @param height The height, or null if not set
     */
    public void setHeight( Integer height );

    /**
     * Get the width of the Panel in pixels
     * @return The width, or null if not set
     */
    public Integer getWidth();

    /**
     * Set the width of the Panel in pixels
     * @param width The width, or null if not set
     */
    public void setWidth( Integer width );

    /**
     * Get the minimum height of the Panel in pixels
     * @return The minimum height, or null if not set
     */
    public Integer getMinHeight();

    /**
     * Set the minimum height of the Panel in pixels
     * @param minHeight The minimum height, or null if not set
     */
    public void setMinHeight( Integer minHeight );

    /**
     * Get the minimum width of the Panel in pixels
     * @return The minimum width, or null if not set
     */
    public Integer getMinWidth();

    /**
     * Set the minimum width of the Panel in pixels
     * @param minWidth The width, or null if not set
     */
    public void setMinWidth( Integer minWidth );

    /**
     * Get the Position of the Panel relate to it's Parent
     * @return The Position of the Panel
     */
    public Position getPosition();

    /**
     * Set the Position of the Panel relative to it's parent.
     * @param position The Position of the Panel relative to it's parent
     */
    public void setPosition( Position position );

    /**
     * Has the Panel been collapsed to a minimal size
     * @return true If minimized
     */
    public boolean isMinimized();

    /**
     * Has the Panel been expanded to a maximum size
     * @return true If maximized
     */
    public boolean isMaximized();

    void setContextDefinition( final ContextDefinition contextDefinition );

    ContextDefinition getContextDefinition();

    ContextDisplayMode getContextDisplayMode();

    void setContextDisplayMode( final ContextDisplayMode contextDisplayMode );

}
