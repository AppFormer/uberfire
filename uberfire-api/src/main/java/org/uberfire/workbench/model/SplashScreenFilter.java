package org.uberfire.workbench.model;

import java.util.Collection;

import org.uberfire.mvp.PlaceRequest;

public interface SplashScreenFilter {

    String getName();

    void setName( final String name );

    boolean displayNextTime();

    void setDisplayNextTime( final boolean value );

    Collection<String> getInterceptionPoints();

    void setInterceptionPoints( final Collection<String> places );

}
