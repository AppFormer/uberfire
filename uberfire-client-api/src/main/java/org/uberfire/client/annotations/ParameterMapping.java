package org.uberfire.client.annotations;

import java.lang.annotation.ElementType;
import java.lang.annotation.Inherited;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Inherited
@Retention(RetentionPolicy.RUNTIME)
@Target({ ElementType.FIELD})
public @interface ParameterMapping {

    String name();
    String val();

}
