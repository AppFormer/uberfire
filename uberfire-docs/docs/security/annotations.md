# Security Annotations

Two security annotations are provided as an alternative to the security checks methods provided by the _AuthorizationManager_ interface.

* **@ResourceCheck**

```
import static org.uberfire.shared.authz.ProjectConstants.*;
...

@ResourceCheck(type=PROJECT, action=CREATE)
public void newProject() {
    newProjectPresenter.show( this );
}
```

Every time the method _newProject_ is called the security check defined in the annotation is processed, causing the method body to be executed only if the permission is granted.


* **@PermsissionCheck**

Annotation checks are supported for plain permissions as well.

```
@PermissionCheck("myapp.reports.sales")
public void showSalesReport() {
    salesReportsPresenter.show();
}
```

#### Callbacks

Both annotations provide callback support:

```
@ResourceCheck(type=PROJECT, action=CREATE,
                onGranted="onCreateGranted",
                onDenied="onCreateDenied")
public void newProject() {
    newProjectPresenter.show();
}

public void onCreateGranted() {
    GWT.log("Project creation allowed");
}

public void onCreateDenied() {
    GWT.log("Project creation NOT allowed");
}
```

Callbacks can be used to add extra processing logic.

The example above without annotations:

```
@Inject
AuthorizationManager authzManager;
...

public void newProject() {
    if (authzManager.authorize(PROJECT, CREATE)) {
        newProjectPresenter.show();
        onCreateGranted();
    }
    else {
        onCreateDenied();
    }
}
...
```

#### Restrictions

The security annotations are can only be used under certain conditions:

* Only work on client side code
* Only work on CDI managed beans
* Only can be added to methods that return "void"
* Only work if the method is called from an external component.
* For inner component calls the solution is to use a delegate or a "self" instance.


