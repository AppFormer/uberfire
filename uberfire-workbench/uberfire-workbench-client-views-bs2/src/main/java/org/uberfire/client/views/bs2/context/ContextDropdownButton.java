package org.uberfire.client.views.bs2.context;

import com.github.gwtbootstrap.client.ui.Button;
import com.github.gwtbootstrap.client.ui.DropdownButton;
import com.github.gwtbootstrap.client.ui.base.IconAnchor;
import com.github.gwtbootstrap.client.ui.constants.BaseIconType;
import com.github.gwtbootstrap.client.ui.constants.ButtonType;
import com.github.gwtbootstrap.client.ui.constants.IconPosition;
import com.github.gwtbootstrap.client.ui.constants.IconSize;
import com.github.gwtbootstrap.client.ui.resources.ButtonSize;
import com.google.gwt.event.dom.client.ClickHandler;
import com.google.gwt.event.shared.HandlerRegistration;

public class ContextDropdownButton extends DropdownButton {

    private Button trigger;

    /**
     * Creates a DropdownButton without a caption.
     */
    public ContextDropdownButton() {
        super();
    }

    /**
     * Creates a DropdownButton with the given caption.
     * @param caption the button's caption
     */
    public ContextDropdownButton( String caption ) {
        super( caption );
    }

    public void displayCaret( boolean value ) {
        trigger.setCaret( value );
    }

    /**
     * {@inheritDoc}
     */
    @Override
    protected IconAnchor createTrigger() {
        trigger = new Button();
        trigger.setCaret( true );
        return trigger;
    }

    /**
     * Sets the button's size.
     * @param size the button's size
     */
    public void setSize( ButtonSize size ) {
        trigger.setSize( size );
    }

    /**
     * Sets the button's type.
     * @param type the button's type
     */
    public void setType( ButtonType type ) {
        trigger.setType( type );
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void setBaseIcon( BaseIconType type ) {
        trigger.setBaseIcon( type );
    }

    @Override
    public HandlerRegistration addClickHandler( ClickHandler handler ) {
        return trigger.addClickHandler( handler );
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void setIconSize( IconSize size ) {
        trigger.setIconSize( size );
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void setCustomIconStyle( String customIconStyle ) {
        trigger.setCustomIconStyle( customIconStyle );
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void setIconPosition( IconPosition position ) {
        trigger.setIconPosition( position );
    }
}
