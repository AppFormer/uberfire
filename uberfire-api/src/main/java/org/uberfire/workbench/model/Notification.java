package org.uberfire.workbench.model;

import java.util.Date;

import org.jboss.errai.common.client.api.Assert;
import org.jboss.errai.common.client.api.annotations.MapsTo;
import org.jboss.errai.common.client.api.annotations.Portable;
import org.uberfire.workbench.events.NotificationEvent;
import org.uberfire.workbench.events.NotificationEvent.NotificationType;

/**
 * Represents a notification that has been displayed in the UberFire workbench.
 * <p>
 * Instances of this class are normally created by the Workbench Notification Manager in response to a
 * {@link NotificationEvent} being fired as a CDI event.
 *
 * @see NotificationEvent
 */
@Portable
public class Notification {

    public enum State {
        /**
         * The notification is newly created and has not yet been acknowledged by the user.
         */
        NEW,

        /**
         * The notification has been acknowledged by the user.
         */
        ACKNOWLEDGED
    }

    private NotificationType type;
    private String message;
    private Date timestamp;
    private State state;

    public Notification( @MapsTo("type") NotificationType type,
                         @MapsTo("message") String message,
                         @MapsTo("timestamp") Date timestamp,
                         @MapsTo("state") State state ) {
        this.type = Assert.notNull( type );
        this.message = Assert.notNull( message );
        this.timestamp = Assert.notNull( timestamp );
        this.state = Assert.notNull( state );
    }

    public NotificationType getType() {
        return type;
    }

    public void setType( NotificationType type ) {
        this.type = type;
    }

    public String getMessage() {
        return message;
    }

    public void setMessage( String message ) {
        this.message = message;
    }

    public Date getTimestamp() {
        return timestamp;
    }

    public void setTimestamp( Date timestamp ) {
        this.timestamp = timestamp;
    }

    public State getState() {
        return state;
    }

    public void setState( State state ) {
        this.state = state;
    }

    @Override
    public String toString() {
        return "Notification [type=" + type + ", message=" + message + ", timestamp=" + timestamp + ", state=" + state + "]";
    }
}
