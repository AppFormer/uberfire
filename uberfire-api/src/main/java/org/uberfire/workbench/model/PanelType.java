package org.uberfire.workbench.model;

import org.jboss.errai.common.client.api.annotations.Portable;

@Portable
public enum PanelType {
    ROOT_TAB,
    ROOT_LIST,
    ROOT_SIMPLE,
    ROOT_STATIC,
    SIMPLE,
    SIMPLE_DND,
    MULTI_TAB,
    MULTI_LIST,
    STATIC,
    TEMPLATE
}
