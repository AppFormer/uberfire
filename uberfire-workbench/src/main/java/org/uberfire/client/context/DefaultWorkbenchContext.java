/*
 * Copyright 2012 JBoss Inc
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.uberfire.client.context;

import javax.enterprise.context.ApplicationScoped;
import javax.enterprise.event.Observes;

import org.uberfire.backend.organizationalunit.OrganizationalUnit;
import org.uberfire.backend.repositories.Repository;
import org.uberfire.backend.vfs.Path;
import org.uberfire.client.workbench.context.WorkbenchContext;
import org.uberfire.workbench.events.OrganizationalUnitChangeEvent;
import org.uberfire.workbench.events.PanelFocusEvent;
import org.uberfire.workbench.events.PathChangeEvent;
import org.uberfire.workbench.events.RepositoryChangeEvent;
import org.uberfire.workbench.model.PanelDefinition;

/**
 * Container for the context of the Workbench
 */
@ApplicationScoped
public class DefaultWorkbenchContext implements WorkbenchContext {

    private OrganizationalUnit activeOrganizationalUnit;
    private Repository activeRepository;
    private Path activePath;
    private PanelDefinition activePanel;

    public void setActiveOrganizationalUnit( @Observes final OrganizationalUnitChangeEvent event ) {
        final OrganizationalUnit activeOrganizationalUnit = event.getOrganizationalUnit();
        setActiveOrganizationalUnit( activeOrganizationalUnit );
        setActiveRepository( (Repository) null );
        setActivePath( (Path) null );
    }

    public void setActiveRepository( @Observes final RepositoryChangeEvent event ) {
        final Repository activeRepository = event.getRepository();
        setActiveRepository( activeRepository );
        setActivePath( (Path) null );
    }

    public void setActivePath( @Observes final PathChangeEvent event ) {
        final Path activePath = event.getPath();
        setActivePath( activePath );
    }

    private void setActiveOrganizationalUnit( final OrganizationalUnit activeOrganizationalUnit ) {
        this.activeOrganizationalUnit = activeOrganizationalUnit;
    }

    @Override
    public OrganizationalUnit getActiveOrganizationalUnit() {
        return this.activeOrganizationalUnit;
    }

    private void setActiveRepository( final Repository activeRepository ) {
        this.activeRepository = activeRepository;
    }

    @Override
    public Repository getActiveRepository() {
        return this.activeRepository;
    }

    private void setActivePath( final Path activePath ) {
        this.activePath = activePath;
    }

    @Override
    public Path getActivePath() {
        return this.activePath;
    }

    @Override
    public PanelDefinition getActivePanel() {
        return activePanel;
    }

    private void setActivePanel( @Observes final PanelFocusEvent panelFocusEvent ) {
        this.activePanel = panelFocusEvent.getPanel();
    }
}
