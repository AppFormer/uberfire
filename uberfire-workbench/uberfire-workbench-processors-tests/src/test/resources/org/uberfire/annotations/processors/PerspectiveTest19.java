package org.uberfire.annotations.processors;

import javax.annotation.PostConstruct;
import javax.enterprise.context.ApplicationScoped;

import com.google.gwt.user.client.ui.FlowPanel;
import org.uberfire.client.annotations.ParameterMapping;
import org.uberfire.client.annotations.WorkbenchPanel;
import org.uberfire.client.annotations.WorkbenchPart;
import org.uberfire.client.annotations.WorkbenchParts;
import org.uberfire.client.annotations.WorkbenchPerspective;
import org.uberfire.workbench.model.PanelType;
@ApplicationScoped
@WorkbenchPerspective(
        identifier = "HomePerspective",
        isDefault = true, isTemplate = true)
public class PerspectiveTest19 {



    @WorkbenchPart(part = "noParameterScreen")
    Object nopParameter = new FlowPanel();

    @WorkbenchPanel(isDefault = true)
    @WorkbenchPart(part = "oneParameterScreen", parameters = @ParameterMapping(name="uber", val="fire"))
    Object oneParameter = new FlowPanel();

    @WorkbenchPart(part = "twoParametersScreen", parameters = {@ParameterMapping(name="uber", val="fire"),@ParameterMapping(name="uber1", val="fire1")})
    Object twoParameters = new FlowPanel();

    @PostConstruct
    public void setup() {

    }


}