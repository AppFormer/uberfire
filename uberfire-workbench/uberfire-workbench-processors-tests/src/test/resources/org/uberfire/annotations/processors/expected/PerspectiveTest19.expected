/*
 * Copyright 2012 JBoss Inc
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.uberfire.annotations.processors;

import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import javax.annotation.Generated;
import javax.enterprise.context.Dependent;
import javax.inject.Inject;

import javax.inject.Named;
import org.uberfire.workbench.model.PerspectiveDefinition;
import java.util.HashMap;
import java.util.Map;
import org.uberfire.client.annotations.Perspective;
import org.uberfire.client.mvp.AbstractTemplateWorkbenchPerspectiveActivity;
import org.uberfire.client.mvp.PlaceManager;

import org.uberfire.mvp.PlaceRequest;

import com.google.gwt.user.client.ui.Widget;
import org.uberfire.client.workbench.TemplatePanelDefinitionImpl;
import org.uberfire.client.workbench.TemplatePerspectiveDefinitionImpl;
import org.uberfire.mvp.impl.DefaultPlaceRequest;
import org.uberfire.workbench.model.PanelDefinition;
import org.uberfire.workbench.model.PanelType;
import org.uberfire.workbench.model.Position;
import org.uberfire.workbench.model.impl.PartDefinitionImpl;
@Dependent
@Generated("org.uberfire.annotations.processors.WorkbenchPerspectiveProcessor")
@Named("HomePerspective")
/*
 * WARNING! This class is generated. Do not modify.
 */
public class PerspectiveTest19Activity extends  AbstractTemplateWorkbenchPerspectiveActivity  {

    private static final Collection<String> ROLES = Collections.emptyList();

    private static final Collection<String> TRAITS = Collections.emptyList();

    @Inject
    private PerspectiveTest19 realPresenter;

    @Inject
    //Constructor injection for testing
    public PerspectiveTest19Activity(final PlaceManager placeManager) {
        super( placeManager );
    }

    @Override
    public String getIdentifier() {
        return "HomePerspective";
    }

    @Override
    public boolean isDefault() {
        return true;
    }

    @Override
    public Collection<String> getRoles() {
        return ROLES;
    }

    @Override
    public Collection<String> getTraits() {
        return TRAITS;
    }

    @Override
    public String getSignatureId() {
        return "org.uberfire.annotations.processors.PerspectiveTest19Activity";
    }
    @Override
    public Widget getRealPresenterWidget( ) {
      return realPresenter.asWidget();
    }

    @Override
    public void setWidget( String fieldName,
        Widget widget ) {

        if ( fieldName.equalsIgnoreCase( "oneParameter" ) ) {
            realPresenter.oneParameter.clear();
            realPresenter.oneParameter.add( widget.asWidget() );
        }
        if ( fieldName.equalsIgnoreCase( "nopParameter" ) ) {
            realPresenter.nopParameter.clear();
            realPresenter.nopParameter.add( widget.asWidget() );
        }
        if ( fieldName.equalsIgnoreCase( "twoParameters" ) ) {
            realPresenter.twoParameters.clear();
            realPresenter.twoParameters.add( widget.asWidget() );
        }
    }

    @Perspective
    public PerspectiveDefinition getPerspective() {
        final PerspectiveDefinition p = new TemplatePerspectiveDefinitionImpl( this,"oneParameter", getClass().getName() );
        PanelDefinition panelDefinition = new TemplatePanelDefinitionImpl( this, PanelType.TEMPLATE , "oneParameter"  );
            Map properties = new HashMap<String,String>();
            properties.put("uber","fire");
            panelDefinition.addPart(
            new PartDefinitionImpl(new DefaultPlaceRequest( "oneParameterScreen", properties  ) ) );
        p.getRoot().appendChild( panelDefinition );

        PanelDefinition panelDefinition0 = new TemplatePanelDefinitionImpl( this, PanelType.TEMPLATE , "nopParameter"  );
        panelDefinition0.addPart(
                new PartDefinitionImpl(new DefaultPlaceRequest( "noParameterScreen" ) ) );
        p.getRoot().appendChild( panelDefinition0 );
        PanelDefinition panelDefinition1 = new TemplatePanelDefinitionImpl( this, PanelType.TEMPLATE , "twoParameters"  );
        Map properties1 = new HashMap<String,String>();
            properties1.put("uber","fire");
            properties1.put("uber1","fire1");
        panelDefinition1.addPart(
                new PartDefinitionImpl(new DefaultPlaceRequest( "twoParametersScreen", properties1 ) ) );
        p.getRoot().appendChild( panelDefinition1 );
        return p;
    }

}
